x/perp/keeper/amm.go

```diff
@@ -27,30 +27,91 @@ func (k Keeper) EditPoolPegMultiplier(
 	if err != nil {
 		return
 	}
+	costInt := cost.Ceil().TruncateInt()
+
+	err = k.handleMarketUpdateCost(ctx, pair, costInt)
+	if err != nil {
+		return
+	}
+
+	// Do the re-peg
+	err = k.PerpAmmKeeper.EditPoolPegMultiplier(ctx, pair, pegMultiplier)
+	if err != nil {
+		return
+	}
 
+	_ = ctx.EventManager().EmitTypedEvent(&types.PegMultiplierUpdate{ // nolint:errcheck
+		Pair:   pair,
+		NewPeg: pegMultiplier,
+		Cost:   costInt,
+	})
+
+	return
+}
+
+func (k Keeper) EditPoolSwapInvariant(ctx sdk.Context, sender sdk.AccAddress, pair asset.Pair, multiplier sdk.Dec) (err error) {
+	// Get the pool
+	pool, err := k.PerpAmmKeeper.GetPool(ctx, pair)
+	if nil != err {
+		return
+	}
+
+	// Compute cost of re-pegging the pool
+	cost, err := pool.GetSwapInvariantUpdateCost(multiplier)
+	if err != nil {
+		return
+	}
 	costInt := cost.Ceil().TruncateInt()
 
-	if costInt.IsPositive() {
+	err = k.handleMarketUpdateCost(ctx, pair, costInt)
+	if err != nil {
+		return
+	}
+
+	// Do the swap invariant update
+	err = k.PerpAmmKeeper.EditSwapInvariant(ctx, pair, multiplier)
+	if err != nil {
+		return
+	}
+
+	// Get the pool
+	pool, err = k.PerpAmmKeeper.GetPool(ctx, pair)
+	if nil != err {
+		return
+	}
```

Comment:
It looks like this is an unnecessary duplicate read since you already have the pool value in a variable